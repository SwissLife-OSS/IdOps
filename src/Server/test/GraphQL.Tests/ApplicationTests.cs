using System;
using System.Collections.Generic;
using System.Linq;
using System.Threading.Tasks;
using HotChocolate;
using HotChocolate.Execution;
using IdOps.Authorization;
using IdOps.Security;
using Snapshooter;
using Snapshooter.Xunit;
using Squadron;
using Xunit;

namespace IdOps.GraphQL.Tests
{
    [Collection(TestCollectionNames.GraphQL)]
    public class ApplicationTests
        : TestHelper
    {
        public ApplicationTests(MongoResource resource) : base(resource)
        {
        }

        [Theory]
        [InlineData(Roles.Admin)]
        [InlineData(Roles.Edit)]
        public async Task ValidRequest(string role)
        {
            // arrange
            ITestRequestBuilder requestBuilder = TestRequestBuilder.New()
                .AddExecutor(await CreateSchema())
                .AddRequestFromFile("CreateApplication")
                .AddUser()
                .AddRole(role);

            await TestDataBuilder
                .New(Services)
                .SetupTenant()
                .ExecuteAsync();

            // act
            IExecutionResult result = await requestBuilder.ExecuteAsync();

            // assert
            result.ToJson().MatchSnapshot(x => x.IgnoreFields("**.id"));
        }

        [Theory]
        [InlineData(null)]
        [InlineData(Roles.Read)]
        public async Task CreateApplication_ByUserWithoutPermission_IsDenied(string? role)
        {
            // arrange
            ITestRequestBuilder requestBuilder = TestRequestBuilder.New()
                .AddExecutor(await CreateSchema())
                .AddUser()
                .AddRole(role)
                .AddRequestFromFile("CreateApplication");
            
            await TestDataBuilder
                .New(Services)
                .SetupTenant()
                .ExecuteAsync();

            // act
            IExecutionResult result = await requestBuilder.ExecuteAsync();

            // assert
            result.ToJson().MatchSnapshot();
        }

        [Fact]
        public async Task CreateApplication_NotAuthenticated()
        {
            // arrange
            ITestRequestBuilder requestBuilder = TestRequestBuilder.New()
                .AddExecutor(await CreateSchema())
                .AddRequestFromFile("CreateApplication")
                .SetAuthenticated(false);
            
            await TestDataBuilder
                .New(Services)
                .SetupTenant()
                .ExecuteAsync();

            // act
            IExecutionResult result = await requestBuilder.ExecuteAsync();

            // assert
            result.ToJson().MatchSnapshot();
        }
    }
}
