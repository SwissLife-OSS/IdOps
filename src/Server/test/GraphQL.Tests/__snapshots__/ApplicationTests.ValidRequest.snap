{
  "data": {
    "createApplication": {
      "application": {
        "id": "379b5561cf554cc0b2f161260b7ccf10",
        "name": "MyApplication",
        "tenant": "TestTenant",
        "templateId": "4a74d535d5c34e21b5b4137c27b2e6dc",
        "allowedGrantTypes": [
          "grantType"
        ],
        "redirectUris": [
          "https://localhost"
        ],
        "identityScopes": [
          "8cab598b78c34532a4257ae376abe5eb",
          "5404b6fd223e435da28b9da695d7d6f6"
        ],
        "apiScopes": [
          "39da2d550adc4b468ae0c9f80dc6b356"
        ]
      }
    }
  }
}
